### ISSUE TITLE
BuildKite CI: Deploying buildkite agents to a cluster v0

### ISSUE THREAD

**Issue by bkase** [CodaProtocol/coda#4802]
We need to build terraform + helm infrastructure to deploy BuildKite agents to GKE.

Getting just this part to work unlocks more parallelizable work.

Epic: #4762

**Comment by 0x0I** [comment]
Seems as though we should be able to leverage the official Buildkite Helm [chart](https://github.com/buildkite/charts) for now and avoid having to develop our own.

With this in mind, the additional infrastructure code structure should resemble the following:

# Code Structure
```
coda-automation
├── helm
├── scripts
├── services
└── terraform
        ├── infrastructure
        ├── modules
                └── kubernetes
                        └── buildkite-agent
        └── buildkite-clusters
                 └── buildkite-product(product CI/CD)
                 └── (e.g. buildkite-infra, buildkite-data, ...) 
```

**Comment by 0x0I** [comment]
# Test Infrastructure

@yourbuddyconner @nholland94 - Considering the technical stack for the proposed Buildkite agent cluster infra consists of:
- Docker (container-engine)
- Kubernetes (provisioner/orchestrator) + Helm (Kubernetes package manager)
- Terraform (IaC)

We should be able to enable CI/CD pipeline of this repository based on:
- [minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/) + [helm](https://helm.sh/)
- `kubectl`'s [config use-context](https://jamesdefabia.github.io/docs/user-guide/kubectl/kubectl_config_use-context/) functionality
- [kgoss](https://github.com/aelsabbahy/goss/tree/master/extras/kgoss) - a wrapper for [goss](https://github.com/aelsabbahy/goss) that aims to bring the simplicity of testing with goss to containers running in pods in Kubernetes
- organization-wide or dedicated infrastructure Buildkite cluster

**Comment by yourbuddyconner** [comment]
I love the use of goss here, and kgoss is especially neat. Overall the plan LGTM,

**Comment by 0x0I** [comment]
@yourbuddyconner  fyi: testing of infrastructure bits/components using `kgoss` and by extension `minikube` is currently blocked by *minikube-issue* [#7420](https://github.com/kubernetes/minikube/issues/7420).

**tl;dr:** within minikube, user override of docker-daemon related environment variables such as (`DOCKER_HOST=`, which is used in the **podsidecar DinD** setup) is ignored as the variables are reset to defaults meant to target only the host docker daemon (`unix:///var/run/docker.sock`).

Discussion is ongoing *re:* resolving this by hopefully not resetting overrides and trusting the developer/operator/user to do the right thing, especially in a test context.

**Comment by 0x0I** [comment]
(**re:** above `minikube` and `DOCKER_HOST` issue above) *Tracking* issue: https://github.com/kubernetes/minikube/issues/8219

**Comment by 0x0I** [comment]
**FYI:** *CircleCI* resource classes/sizing (**re:** Buildkite agent cluster construction)

![Screenshot from 2020-05-27 10-01-40](https://user-images.githubusercontent.com/49376577/83030356-85e4d780-a001-11ea-9a9a-92c27a1c601a.png)

**Comment by 0x0I** [comment]
**v0** of cluster has been provisioned and is operational.

*Note:* test infrastructure design and implementation will be moved to the **v1** cluster iteration.
